File: problems/LeetCode/tree/sumNumbers.py
def sumNumbers1(self, root: Optional[TreeNode]) -> int:

		ans = 0

		def dfs(node: Optional[TreeNode], currentSum: str) -> None:
				if not node:
						return
				
				currentSum += str(node.val)

				if not node.left and not node.right:
						nonlocal ans
						ans += int(currentSum)
						return
				
				dfs(node.left, currentSum)
				dfs(node.right, currentSum)
		
		dfs(root, '')

		return ans


def sumNumbers(self, root: Optional[TreeNode]) -> int:

		ans = 0

		def dfs(node: Optional[TreeNode], currentSum: int) -> None:
				if not node:
						return
				
				currentSum = currentSum * 10 + node.val

				if not node.left and not node.right:
						nonlocal ans
						ans += currentSum
						return
				
				dfs(node.left, currentSum)
				dfs(node.right, currentSum)
		
		dfs(root, 0)

		return ans

File: problems/LeetCode/TopInterview150/sumNumbers.ts
function dfs(root: TreeNode | null, str: string, total: { sum: number }) {
  if (!root) {
    return;
  }

  str += String(root.val);

  if (!root.left && !root.right) {
    total.sum += Number(str);
  }

  dfs(root.left, str, total);
  dfs(root.right, str, total);
}

function sumNumbers(root: TreeNode | null): number {
  let total = { sum: 0 };
  dfs(root, '', total);
  return total.sum;
}

function sumNumbers2(root: TreeNode | null): number {
  let currentSum = 0;

  const dfs = (root: TreeNode | null, str: string) => {
    if (!root) {
      return;
    }

    str += String(root.val);

    if (!root.left && !root.right) {
      currentSum += Number(str);
    }

    dfs(root.left, str);
    dfs(root.right, str);
  };

  dfs(root, '');

  return currentSum;
}

// Time complexity: O(n)
// Space compleixty: O(h)
